File: LeetCode/src/stack/ValidParenthesisString.java
package stack;

import java.util.Stack;

public class ValidParenthesisString {

    public boolean checkValidString(String s) {

        Stack<Integer> brackets = new Stack<>();
        Stack<Integer> asterisks = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(') {
                brackets.push(i);
            } else if (c == '*') {
                asterisks.push(i);
            } else {
                if (!brackets.isEmpty()){
                    brackets.pop();
                } else if (!asterisks.isEmpty()) {
                    asterisks.pop();
                } else {
                    return false;
                }
            }
        }

        while (!brackets.isEmpty() && !asterisks.isEmpty()){
            int b = brackets.pop();
            int a = asterisks.pop();
            if (b > a) {
                return false;
            }
        }

        return brackets.isEmpty();

    }

    public static void main(String[] args) {
        ValidParenthesisString solution = new ValidParenthesisString();

        // Test cases
        String s1 = "()";
        String s2 = "(*)";
        String s3 = "(*))";
        String s4 = "(((*)";

        System.out.println("Test Case 1: " + solution.checkValidString(s1)); // Output: true
        System.out.println("Test Case 2: " + solution.checkValidString(s2)); // Output: true
        System.out.println("Test Case 3: " + solution.checkValidString(s3)); // Output: true
        System.out.println("Test Case 4: " + solution.checkValidString(s4)); // Output: false
    }
}

File: LeetCode_Python/stack/ValidParenthesisString.py
class ValidParenthesisString:
    def checkValidString(self, s: str) -> bool:
        left_stack = []
        star_stack = []

        for i, ch in enumerate(s):
            if ch == "(":
                left_stack.append(i)
            elif ch == "*":
                star_stack.append(i)
            else:
                if left_stack:
                    left_stack.pop()
                elif star_stack:
                    star_stack.pop()
                else:
                    return False

        while left_stack and star_stack:
            if left_stack[-1] < star_stack[-1]:
                left_stack.pop()
                star_stack.pop()
            else:
                return False

        return not left_stack


if __name__ == "__main__":
    solution = ValidParenthesisString()

    print("Expected Output: True")
    print("Actual Output:", solution.checkValidString("()"))

    print("Expected Output: True")
    print("Actual Output:", solution.checkValidString("(*)"))

    print("Expected Output: True")
    print("Actual Output:", solution.checkValidString("(*))"))

    print("Expected Output: False")
    print("Actual Output:", solution.checkValidString("(*)))"))
